modules/i3s/src/lib/parsers/parse-i3s-tile-content.js

```diff
@@ -333,3 +335,86 @@ function offsetsToCartesians(vertices, metadata = {}, cartographicOrigin) {
 
   return positions;
 }
+
+/**
+ * Makes PBR material from material definition
+ * @returns {object}
+ */
+function makePbrMaterial(materialDefinition, texture) {
+  if (!materialDefinition) {
+    return null;
+  }
+  const material = getObjectDeepCopy(materialDefinition);
+
+  // Convert colors from [255,255,255,255] to [1,1,1,1]
+  if (material.emissiveFactor) {
+    convertColorFormat(material.emissiveFactor);
+  }
+  if (material.pbrMetallicRoughness && material.pbrMetallicRoughness.baseColorFactor) {
+    convertColorFormat(material.pbrMetallicRoughness.baseColorFactor);
+  }
+
+  setMaterialTexture(material, texture);
+
+  return material;
```

Comment:
You could just make sure you copy any object you modify rather than copying everything

```suggestion
  const material = {...materialDefinition};

  // Convert colors from [255,255,255,255] to [1,1,1,1]
  if (material.emissiveFactor) {
    material.emissiveFactor = convertColorFormat(material.emissiveFactor);
  }
  if (material.pbrMetallicRoughness && material.pbrMetallicRoughness.baseColorFactor) {
    material.pbrMetallicRoughness = {...i3sMaterial.pbrMetallicRoughness);
    material.pbrMetallicRoughness.baseColorFactor convertColorFormat(i3sMaterial.pbrMetallicRoughness.baseColorFactor);
  }

  setMaterialTexture(material, texture);

  return material;
```

Reply:
I made it deep-copy because source object is `materialDefinition` https://github.com/Esri/i3s-spec/blob/master/docs/1.7/materialDefinitions.cmn.md and it is used for many nodes. We must copy not only this `materialDefinition` but all embedded objects: `pbrMetallicRoughness` and textures. If we don't do it all that objects mutate.

Reply:
My point was that we should never mutate these objects. That is why we only need to copy the ones we are changing, but we can share "trees" with the others. 

Small object creation can really hurt performance.

Reply:
I see. In that case we at least need to copy `pbrMetallicRougness`